x/leverage/keeper/interest.go

```diff
@@ -172,3 +182,29 @@ func (k *Keeper) InitializeLastInterestTime(ctx sdk.Context) {
 		store.Set(timeKey, bz)
 	}
 }
+
+// initializeBorrowApyMap initiate the borrowApy map with zeroDec of all tokens registered
+func (k *Keeper) initializeBorrowApyMap(ctx sdk.Context) (map[string]sdk.Dec, error) {
+	borrowApy := map[string]sdk.Dec{}
+
+	allTokens, err := k.GetAllRegisteredTokens(ctx)
+	if err != nil {
+		return borrowApy, err
+	}
+
+	for _, token := range allTokens {
+		borrowApy[token.BaseDenom] = sdk.ZeroDec()
+	}
+
+	return borrowApy, nil
+}
+
+// storeBorrowApy stores all the borrow APY value into the store
+func (k *Keeper) storeBorrowApy(ctx sdk.Context, borrowApy map[string]sdk.Dec) error {
```

Comment:
So this is a very important point. Iterating over a map is non-deterministic and since we're updating state while iterating over a map here, this would result in non-determinism.

However, the SDK actually somewhat protects against this, but I'd still like to instill good practice here and not mutate state while iterating over maps.

We can still use a map, but instead:

1. take the keys and store them into a slice
2. sort the slice via `sort.Strings`
3. iterate over the slice and then mutate state.

Reply:
Nice, I didn't know about that 
One idea to avoid storing the keys into a slice and then iterate over them would be during the  `range totalBorrowed `  like this 
```go
	for _, coin := range totalBorrowed {
		reserveFactor, err := k.GetReserveFactor(ctx, coin.Denom)
		if err != nil {
			return err
		}

		utilization, err := k.GetBorrowUtilization(ctx, coin.Denom, coin.Amount)
		if err != nil {
			return err
		}

		derivedRate, err := k.GetDynamicBorrowInterest(ctx, coin.Denom, utilization)
		if err != nil {
			return err
		}

		reserveFactors[coin.Denom] = reserveFactor
		interestToApply[coin.Denom] = derivedRate.Mul(yearsElapsed)

		if err := k.SetBorrowAPY(ctx, coin.Denom, derivedRate); err != nil {
			return err
		}
	}

```

- The good thing is to avoid storing a map or slice
- The drawback is to add complexity to the for loop 

Let me know your thoughts @alexanderbez

Reply:
That looks like the best way to do it

Reply:
ACK